qiskit/algorithms/optimizers/qnspsa.py

```diff
@@ -0,0 +1,245 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A generalized SPSA optimizer including support for Hessians."""
+
+from typing import Iterator, Optional, Union, Callable, Dict
+
+import numpy as np
+from qiskit.providers import Backend
+from qiskit.circuit import ParameterVector, QuantumCircuit
+from qiskit.opflow import StateFn, CircuitSampler, ExpectationBase
+from qiskit.utils import QuantumInstance
+
+from .spsa import SPSA, _batch_evaluate
+
+# the function to compute the fidelity
+FIDELITY = Callable[[np.ndarray, np.ndarray], float]
+
+# parameters, loss, stepsize, number of function evaluations, accepted
+CALLBACK = Callable[[np.ndarray, float, float, int, bool], None]
+
+
+class QNSPSA(SPSA):
+    r"""The Quantum Natural SPSA (QN-SPSA) optimizer.
+
+    The QN-SPSA optimizer [1] is a stochastic optimizer that belongs to the family of gradient
+    descent methods. This optimizer is based on SPSA but attempts to improve the convergence by
+    sampling the **natural gradient** instead of the vanilla, first-order gradient. It achieves
+    this by approximating Hessian of the ``fidelity`` of the ansatz circuit.
+
+    Compared to natural gradients, which require :math:`\mathcal{O}(d^2)` expectation value
+    evaluations for a circuit with :math:`d` parameters, QN-SPSA only requires
+    :math:`\mathcal{O}(1)` and can therefore significantly speed up the natural gradient calculation
+    by sacrificing some accuracy. Compared to SPSA, QN-SPSA requires 4 additional function
+    evaluations of the fidelity.
+
+    The stochastic approximation of the natural gradient can be systematically improved by
+    increasing the number of ``resamplings``. This leads to a Monte Carlo-style convergence to
+    the exact, analytic value.
+
+    Example:
+
+        This short example runs QN-SPSA for the ground state calculation of the ``Z ^ Z``
+        observable where the ansatz is a ``PauliTwoDesign`` circuit. For a real experiment,
+        the number of iterations (here 10) should be increased to several hundred steps.
+
+        .. jupyter-execute::
+
+            import numpy as np
+            from qiskit.algorithms.optimizers import QNSPSA
+            from qiskit.circuit.library import PauliTwoDesign
+            from qiskit.opflow import Z, StateFn
+
+            ansatz = PauliTwoDesign(2, reps=1, seed=2)
+            observable = Z ^ Z
+            initial_point = np.random.random(ansatz.num_parameters)
+
+            def loss(x):
+                bound = ansatz.bind_parameters(x)
+                return np.real((StateFn(observable, is_measurement=True) @ StateFn(bound)).eval())
+
+            fidelity = QNSPSA.get_fidelity(ansatz)
+            qnspsa = QNSPSA(fidelity, maxiter=10)
+            result = qnspsa.optimize(ansatz.num_parameters, loss, initial_point=initial_point)
```

Comment:
Is there a reason to use a jupyter execute here? This adds runtime to the docs build because it's executed during the build, but from what I can tell there is no output actually displayed from the execution. If this is just a code example lets use a code block to avoid unnecessary build time. That or add some generated output (like a visualization or something)  to justify the build time execution.

Reply:
No not at all, it was more to show how to use this. I'll change it to an ``Examples`` block